### PR TITLE
Improve suggestions

### LINKED ISSUE #1052: audit all suggestions for optional parens
1. Create a function that takes a parent-expr, child-expr, suggestion-text and produces a suggestion text with potentially added parens
2. change all existing parens adding suggestions to use the function
3. audit all other suggestions to check if they need to use the function
4. Bonus points: do 3. with a lint.

### PR DESCRIPTION
**Pull request by mcarton**
Fix #1052.
This is a work in progress. I probably have forgotten most suggestions. I'll review them all tomorrow.

### CODE REVIEW

File: tests/compile-fail/float_cmp.rs [hunk 1]
```diff
@@ -40,23 +40,47 @@ fn main() {
     ZERO == 0.0; //no error, comparison with zero is ok
     ZERO + ZERO != 1.0; //no error, comparison with zero is ok
 
-    ONE == 1f32; //~ERROR ==-comparison of f32 or f64
-    ONE == (1.0 + 0.0); //~ERROR ==-comparison of f32 or f64
-
-    ONE + ONE == (ZERO + ONE + ONE); //~ERROR ==-comparison of f32 or f64
```
Comment (mcarton) [inline tests/compile-fail/float_cmp.rs]: Feeling like cheating anyone?

**Comment by mcarton** [comment]
(This is now based on #1054.)

Some more better suggestions. Introducing `multispan_sugg`, a function to build a suggestion made from several replacements, without playing with spans manually.
Still a WIP, there are much more suggestions to review.

**Comment by mcarton** [comment]
Ok, this PR has become big enough. I think it's ready to be merged.

**Comment by oli-obk** [comment]
I'm at ffa840d

File: clippy_lints/src/needless_bool.rs [hunk 2]
```diff
@@ -49,11 +50,20 @@ impl LateLintPass for NeedlessBool {
...
+            let reduce = |ret, not| {
+                let snip = Sugg::hir(cx, pred, "<predicate>");
+                let snip = if not {
+                    !snip
+                } else {
+                    snip
+                };
+
+                let hint = if ret {
+                    format!("return {};", snip)
```
Comment (oli-obk) [inline clippy_lints/src/needless_bool.rs]: The semicolon is wrong here, if the if-then-else is a match arm (or a function argument or sth)

**Comment by Manishearth** [comment]
I've had a broad look and am very much 🎉 🎈 🎆 for this, but haven't had the time for a thorough review, yet. Thanks for doing this, though!

**Comment by oli-obk** [comment]
I'm through. The `Sugg` struct is awesome. A lot of code became more readable. Especially where the ad-hoc parenthesis adders were factored out.

Some parts are a little confusing, like `sugg.as_ty(&format!("{} as {}", x, y))`. I get that the "type" is just a string, so we can dump anything there, and that there's not much advantage over `sugg.as_ty(x.to_string()).as_ty(y.to_string())`, but it feels weird. But it's not a blocker imo.

Except for the needless bool refactoring, I could not find any bugs that were introduced from the refactoring.

**Comment by mcarton** [comment]
> Some parts are a little confusing, like `sugg.as_ty(&format!("{} as {}", x, y))`. I get that the "type" is just a string, so we can dump anything there, and that there's not much advantage over `sugg.as_ty(x.to_string()).as_ty(y.to_string())`, but it feels weird. But it's not a blocker imo.

But `sugg.as_ty(x.to_string()).as_ty(y.to_string())` is 17% longer as `sugg.as_ty(&format!("{} as {}", x, y))`! :sweat_smile: 

> Except for the needless bool refactoring, I could not find any bugs that were introduced from the refactoring.

Let's see what `rustfmt` suggests about that:
```rust
% cat a.rs
fn foo() -> i32 {
    return match 42 {
        1337 => 314,
        _ => 0xdeadbeef,
    }
}
% rustfmt a.rs
% cat a.rs
fn foo() -> i32 {
    return match 42 {
        1337 => 314,
        _ => 0xdeadbeef,
    };
}
% cat b.rs
fn foo() -> i32 {
    return if true {
        42
    } else {
        1337
    }
}
% rustfmt b.rs
% cat b.rs
fn foo() -> i32 {
    return if true {
        42
    } else {
        1337
    };
}
```

I think the convention is to always have a semi-colon with an explicit `return`.

**Comment by Manishearth** [comment]
@nrc Btw, this API might be useful for a rustfix tool in the future.

**Comment by Manishearth** [comment]
_Theoretically_ we could tweak this API so that it can actually fix the code by editing it directly! 😄 

I've always wanted such a mode for rustc and clippy. Still involves some work, but this PR gets most of it done

**Comment by oli-obk** [comment]
I meant `match x { a => return foo, _ => {} }`

**Comment by oli-obk** [comment]
>  is 17% longer

*cough* make the method generic over `Display` *cough*

**Comment by mcarton** [comment]
> cough make the method generic over `Display` cough

Duh me.

> Theoretically we could tweak this API so that it can actually fix the code by editing it directly! :smile:

Needs more works :smile:, but that would be great.

**Comment by Manishearth** [comment]
And now, https://github.com/killercup/rustfix exists!

**Comment by killercup** [comment]
@mcarton, before I open a new issue: Does this change by any chance fix that some suggestions for lines that end with `;` do not currently end with a `;`?

Just got this one in my rustfix demo:

```plain
Info: transmute from a pointer type (`*mut libc::c_void`) to a reference type (`&mut Box<for<'r, 'r> std::ops::FnMut(&'r menus::MenuItem, &'r windows::Window)>`)
      #[warn(transmute_ptr_to_ref)] on by default
  --> src/menus.rs:50:17-51:76
Suggestion - Replace:

    mem::transmute::<*mut c_void,
                     &mut Box<FnMut(&MenuItem, &Window)>>(data)(&menu_item, &window);

with:

    &mut *(data as *mut Box<for<'r, 'r> std::ops::FnMut(&'r menus::MenuItem, &'r windows::Window)>)
```

(not the missing semicolon in the suggestion).

If not, feel free to copy this comment into a new issue 😄

**Comment by mcarton** [comment]
@killercup not yet, the context isn't used. But it should. I've been planning to do a second round with that.

**Comment by killercup** [comment]
@mcarton Okay, cool. I'll just paste that into a new issue then :)

Already worked around this with an [ugly patch](https://github.com/killercup/rustfix/commit/ee128f315427c6ea2e0091b3b8fc66545246b51d) anyway 😄

**Comment by mcarton** [comment]
@killercup the thing is not all suggestions might need a `;`. Imagine the `needless_ret` might suggestion to replace `return x;` with `x`.

**Comment by killercup** [comment]
@mcarton Right, I totally missed that. Are there other lints that might do that?

I opened #1076, btw.

**Comment by mcarton** [comment]
I'm not sure. For that specific example probably not many.
Don't hesitate to report other things like that! I'd really want a “just fix it” tool :tada: :sparkles:

**Comment by llogiq** [comment]
So let's merge this already!

**Pull request merged**